test/e2e/cloud-provider-azure.go

```diff
@@ -23,122 +23,54 @@ import (
-	}, input.WaitForControlPlaneIntervals...).Should(Succeed())
+	values := []string{fmt.Sprintf("infra.clusterName=%s", input.ConfigCluster.ClusterName)}
+	InstallHelmChart(ctx, input, cloudProviderAzureHelmRepoURL, cloudProviderAzureChartName, cloudProvierAzureHelmReleaseName, values)
 	By("Waiting for a Running cloud-controller-manager pod")
-	Eventually(func() bool {
-		clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
-		clientSet := clusterProxy.GetClientSet()
-		var runningPods int
-		list, err := clientSet.CoreV1().Pods("kube-system").List(ctx, v1.ListOptions{
-			LabelSelector: "component=cloud-controller-manager",
-		})
-		if err != nil {
-			return false
-		}
-		for _, p := range list.Items {
-			if p.Status.Phase == corev1.PodRunning {
-				runningPods++
-			}
-		}
-		return runningPods > 0
-	}, input.WaitForControlPlaneIntervals...).Should(BeTrue())
-	By("Waiting for Running cloud-node-manager pods")
-	Eventually(func() bool {
-		clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
-		clientSet := clusterProxy.GetClientSet()
-		var runningPods int64
-		list, err := clientSet.CoreV1().Pods("kube-system").List(ctx, v1.ListOptions{
-			LabelSelector: "k8s-app=cloud-node-manager",
-		})
-		if err != nil {
-			return false
-		}
-		for _, p := range list.Items {
-			if p.Status.Phase == corev1.PodRunning {
-				runningPods++
-			}
-		}
-		return runningPods >= to.Int64(input.ConfigCluster.ControlPlaneMachineCount)
-	}, input.WaitForControlPlaneIntervals...).Should(BeTrue())
+	clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
+	workloadClusterClient := clusterProxy.GetClient()
+	cloudControllerManagerPodLabel, err := labels.Parse("component=cloud-controller-manager")
+	Expect(err).ToNot(HaveOccurred())
+	framework.WaitForPodListCondition(ctx, framework.WaitForPodListConditionInput{
+		Lister: workloadClusterClient,
+		ListOptions: &client.ListOptions{
+			LabelSelector: cloudControllerManagerPodLabel,
+			Namespace:     "kube-system",
+		},
+		Condition: podListHasNumPods(1),
+	}, input.WaitForControlPlaneIntervals...)
+	mgmtClusterClient := input.ClusterProxy.GetClient()
+	machineList, err := getMachinesInCluster(ctx, mgmtClusterClient, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
+	Expect(err).ToNot(HaveOccurred())
+	By(fmt.Sprintf("Waiting for %d Running cloud-node-manager pods", len(machineList.Items)))
+	cloudNodeManagerPodLabel, err := labels.Parse("k8s-app=cloud-node-manager")
+	Expect(err).ToNot(HaveOccurred())
+	framework.WaitForPodListCondition(ctx, framework.WaitForPodListConditionInput{
+		Lister: workloadClusterClient,
+		ListOptions: &client.ListOptions{
+			LabelSelector: cloudNodeManagerPodLabel,
+			Namespace:     "kube-system",
+		},
+		Condition: podListHasNumPods(len(machineList.Items)),
```

Comment:
@CecileRobertMichon I think counting the machines in the cluster is probably the best approach for this

Reply:
what about machine pools? Should we count nodes instead?

Reply:
We should probably count both Machines and MachinePoolMachines, right? (a cluster can exist with both types, I assume)

Reply:
IMO doing it by number of nodes is more future-proof. There's no guarantee CAPI won't add a new type of Machine in the future.

Reply:
The problem with using nodes is that they are not the authoritative unit in terms of calculating "expected number" outcomes. For example, a node may be offline, or not yet joined, or other states that are not as definitive in terms of gathering a cluster count as compared to number of machines (plus number of machinepoolmachines).

So, if we use nodes then we'd probably want to change the pod list criteria to >= to account for the fact that a new node may have joined after we calculated num nodes, but before we gathered a list of cloud-node-manager pods.

It would be nice to establish a resilient (hopefully simple to understand and maintain) pattern here, as I do think being able to both validate the helm install and validate expected outcomes of that helm install is valuable.

Reply:
I see. In that case thoughts about adding helper functions to the CAPI test framework that return the number of expected control plane nodes and the number of expected worker nodes? Control planes can be the total number of control plane replicas of the control plane object (or 0 if the control plane does not implement replicas https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html?highlight=spec#required-spec-fields-for-implementations-using-replicas). Worker nodes can be the total of machines + machine pool replicas. 

It's a bit more resilient to abstract what a worker machine can be and have one central place to make that assumption than having to update every place in our CAPZ tests where we make that assumption later if a new type of Machine is added. 

Also worth checking if something like this doesn't already exist in CAPI, it's very possible it does.

Reply:
I checked and capi E2E doesn't have a convenience for this (there are things sort of similar, with lots of buried `Expect()` funcs that are in the process of being cleaned up, but not for this purpose explicitly). How about this as a way of getting this merged ASAP.

1) We verify the existence of the cloud-node-manager daemonset
2) We wait until all of its replicas are Ready